### PR TITLE
Add OCI Generative AI tool calling support

### PR DESCRIPTION
**Pull request by LejianLeoHe7**
# Description

Adding tool calling support for chat models offered by OCI Generative AI services. 
This is an update to our last PR #14080.

Fixes # (issue)
N/A

## New Package?
No

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [X] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [X] Yes
- [ ] No

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- [X] I added new unit tests to cover this change
- [ ] I believe this change is already covered by existing unit tests

## Suggested Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py [hunk 1]
```diff
@@ -219,6 +190,7 @@ def _get_all_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
             **kwargs,
         }
 
+    @deprecated("Deprecated in favor of `chat`, which should be used instead.")
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py:193]: technically, you can make complete call chat -- we have utils to make this very easy. A few llms do this

File: llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py [hunk 2]
```diff
@@ -322,26 +306,164 @@ def gen() -> ChatResponseGen:
...
-    ) -> CompletionResponse:
-        # do synchronous complete for now
-        return self.complete(prompt, formatted=formatted, **kwargs)
-
     async def achat(
         self, messages: Sequence[ChatMessage], **kwargs: Any
     ) -> ChatResponse:
-        # do synchronous chat for now
-        return self.chat(messages, **kwargs)
+        raise NotImplementedError("Async chat is not implemented yet.")
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py:312]: There's no async client methods we can implement?

File: llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py [hunk 3]
```diff
@@ -322,26 +306,164 @@ def gen() -> ChatResponseGen:
...
+        chat_params = self._get_all_kwargs(**kwargs)
+
+        return {
+            "messages": messages,
+            "tools": tool_specs,
+            **oci_params,
+            **chat_params,
+        }
+
+    def chat_with_tools(
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py:356]: Technically you shouldn't need to implement chat_with_tools or stream_chat_with_tools -- the idea is that chat and stream chat already handle tools being passed in/out, and all you need to implement is `_prepare_chat_with_tools()` and `get_tool_calls_from_response()` and maybe `_validate_chat_with_tools_response()`

Check out anthropic or Ollama for a decent example of this

File: llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py (hunk 1 above)
Comment (YouNeedCryDear) [reply llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py:193]: Thank you @logan-markewich Agreed. But basically all the models that support completion have been deprecated in our service. Let me take a look at other vendors.

File: llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py (hunk 2 above)
Comment (YouNeedCryDear) [reply llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py:312]: @logan-markewich Our Python SDK doesn't have async support natively. We probably need to re-evaluation how we gonna implement async. Probably will be another PR as well.

File: llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py (hunk 3 above)
Comment (LejianLeoHe7) [reply llama-index-integrations/llms/llama-index-llms-oci-genai/llama_index/llms/oci_genai/base.py]: Hi @logan-markewich , I have updated chat and stream_chat and also removed chat_with_tools/stream_chat_with_tools in our implementation. We tested both chat/stream_chat with and without tools, which worked fine. Thanks for the suggestions!

**Comment by YouNeedCryDear** [comment]
@logan-markewich Hi, we have addressed all your comments. Please take another look when you have a moment and let us know if there are any remaining issues. We’re eager to get this merged, as our customers are eagerly awaiting this feature.

**Review by YouNeedCryDear — APPROVED** [review]

**Comment by YouNeedCryDear** [comment]
@logan-markewich Hi, could you kindly take a look at the PR when you've got time?

**Review by logan-markewich — APPROVED** [review]

**Comment by logan-markewich** [comment]
I guess poetry lock failed somewhere? I'll take a look

**Comment by logan-markewich** [comment]
Published -- the min python version had to be bumped to 3.9

**Pull request merged**